Editor/AGS.Editor/Validation/ValidationResult.cs

```diff
@@ -0,0 +1,75 @@
+﻿using System;
+using System.Linq;
+using System.Text;
+
+namespace AGS.Editor.Validation
+{
+    // mark as readonly on C# 7.2 or later
+    public struct ValidationResult
```

Comment:
Please do some things differently in the design here.

1. Do you need to use a struct? As a general rule of thumb in C# you should't use structs unless you're sure they're right for the job, since classes and structs are handled differently in memory using them wrong can have penalties. I assume it was because you can mark it as readonly, but you can get the same effect with a class by marking the instances in the class as readonly, which is something you should do anyway.

2. I already mentioned this in 1. but `errors` and `warnings` should be marked readonly

3. Do null checking in the constructor and initialize to empty arrays if null. The properties here are a lot more difficult to interpret because of heavy conditional checking, and they don't need to be this complicated if the instance fields was just guarded from being null in the first place. Anyone who comes in to read the code here after you will be forever grateful for doing this little change.

4. Consider making the properties return `IEnumerable<T>` instead of hard fixed arrays. IEnumerable can be inserted into any collection type in the .NET API with no performance penalty, so they're perfect to be used as public API points. They can also be for eached without converting to a type. If you must have them as a type it would be something like `ValidationResult[] array =  vr.Errors.ToArray()`. Another reason this is important is that it helps to make your type immutable because IEnumerable is exactly that, while arrays can have individual elements overwritten when you reference them from the type, for example `vr.Errors[4] = "this has now been edited"`

Doing something like this will make your type immutable and also much more pleasant to read and work with

```
    public class ValidationResult
    {
        private readonly List<string> _errors = new List<string>();
        private readonly List<string> _warnings = new List<string>();

        public ValidationResult()
        {
        }

        public ValidationResult(IEnumerable<string> errors) : this()
        {
            _errors.AddRange(errors);
        }

        public ValidationResult(IEnumerable<string> errors, IEnumerable<string> warnings) : this(errors)
        {
            _warnings.AddRange(warnings);
        }

        public IEnumerable<string> Errors => _errors;
        ....
     }
```